File: services/rcsi/internal/index/proof.go
package index

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"time"
)

// ValidateProof verifies a proof against a snapshot of the index state.
func ValidateProof(proof Proof, snapshot IndexSnapshot) error {
	if proof.Version > snapshot.Version {
		return fmt.Errorf("proof version %d is ahead of snapshot version %d", proof.Version, snapshot.Version)
	}

	switch proof.Kind {
	case ProofKindDocument:
		if proof.Document == "" {
			return errors.New("document proof missing document id")
		}
		if containsDocument(snapshot.Documents, proof.Document) {
			return fmt.Errorf("document %s still present in snapshot", proof.Document)
		}
		ts, ok := findDocumentTombstone(snapshot.DocumentTombstones, proof.Document)
		if !ok {
			return fmt.Errorf("no tombstone present for document %s", proof.Document)
		}
		expected := digestFor("document", ts.Term, ts.DocumentID, ts.Sequence, mustParseTime(ts.Timestamp), ts.Reason, ts.Version)
		if expected != ts.Digest {
			return fmt.Errorf("digest mismatch for document %s", proof.Document)
		}
	case ProofKindTerm:
		if proof.Term == "" {
			return errors.New("term proof missing term")
		}
		if proof.Document == "" {
			return errors.New("term proof missing document id")
		}
		if postingExists(snapshot.InvertedPostings, proof.Term, proof.Document) {
			return fmt.Errorf("term %s still linked to document %s", proof.Term, proof.Document)
		}
		ts, ok := findTermTombstone(snapshot.TermTombstones, proof.Term, proof.Document)
		if !ok {
			return fmt.Errorf("no term tombstone for %s/%s", proof.Term, proof.Document)
		}
		expected := digestFor("term", proof.Term, proof.Document, ts.Sequence, mustParseTime(ts.Timestamp), ts.Reason, ts.Version)
		if expected != ts.Digest {
			return fmt.Errorf("digest mismatch for term %s and document %s", proof.Term, proof.Document)
		}
	default:
		return fmt.Errorf("unsupported proof kind %s", proof.Kind)
	}

	return nil
}

func digestFor(kind, term, docID string, sequence uint64, ts time.Time, reason string, version uint64) string {
	payload := fmt.Sprintf("%s|%s|%s|%d|%d|%s|%d", kind, term, docID, sequence, ts.UnixNano(), reason, version)
	sum := sha256.Sum256([]byte(payload))
	return hex.EncodeToString(sum[:])
}

func containsDocument(documents []DocumentView, id string) bool {
	for _, doc := range documents {
		if doc.ID == id {
			return true
		}
	}
	return false
}

func postingExists(postings []InvertedPosting, term, docID string) bool {
	for _, posting := range postings {
		if posting.Term != term {
			continue
		}
		for _, candidate := range posting.Documents {
			if candidate == docID {
				return true
			}
		}
	}
	return false
}

func findDocumentTombstone(tombstones []tombstoneView, docID string) (tombstoneView, bool) {
	for _, ts := range tombstones {
		if ts.DocumentID == docID {
			return ts, true
		}
	}
	return tombstoneView{}, false
}

func findTermTombstone(groups []TermTombstoneView, term, docID string) (tombstoneView, bool) {
	for _, group := range groups {
		if group.Term != term {
			continue
		}
		for _, ts := range group.Tombstones {
			if ts.DocumentID == docID {
				return ts, true
			}
		}
	}
	return tombstoneView{}, false
}

func mustParseTime(value string) time.Time {
	ts, err := time.Parse(time.RFC3339Nano, value)
	if err != nil {
		// Fallback to zero value if parsing fails to ensure deterministic behaviour in validation errors.
		return time.Unix(0, 0).UTC()
	}
	return ts
}

File: sdk/typescript/src/rcsi/proof.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.digestFor = digestFor;
exports.validateProof = validateProof;
const crypto_1 = require("crypto");
const NANOS_IN_MILLISECOND = 1000000n;
function timestampToNanoseconds(value) {
    const baseMs = BigInt(Date.parse(value));
    const match = value.match(/\.([0-9]+)(Z|[+-]\d{2}:?\d{2})?$/);
    let nanos = baseMs * NANOS_IN_MILLISECOND;
    if (match && match[1]) {
        const fractional = match[1].padEnd(9, '0').slice(0, 9);
        nanos += BigInt(fractional);
    }
    return nanos;
}
function digestFor(kind, term, documentId, sequence, timestamp, reason, version) {
    const components = [
        kind,
        term ?? '',
        documentId,
        sequence.toString(),
        timestampToNanoseconds(timestamp).toString(),
        reason,
        version.toString(),
    ];
    return (0, crypto_1.createHash)('sha256').update(components.join('|')).digest('hex');
}
function validateProof(proof, snapshot) {
    if (proof.version > snapshot.version) {
        throw new Error(`proof version ${proof.version} is ahead of snapshot ${snapshot.version}`);
    }
    if (proof.kind === 'document') {
        const stillPresent = snapshot.documents.some((doc) => doc.id === proof.documentId);
        if (stillPresent) {
            throw new Error(`document ${proof.documentId} still present in snapshot`);
        }
        const tombstone = snapshot.documentTombstones.find((ts) => ts.documentId === proof.documentId);
        if (!tombstone) {
            throw new Error(`no tombstone for document ${proof.documentId}`);
        }
        const expected = digestFor('document', tombstone.term, tombstone.documentId, tombstone.sequence, tombstone.timestamp, tombstone.reason, tombstone.version);
        if (expected !== tombstone.digest) {
            throw new Error(`digest mismatch for document ${proof.documentId}`);
        }
    }
    else {
        if (!proof.term) {
            throw new Error('term proof requires term');
        }
        const posting = snapshot.inverted.find((entry) => entry.term === proof.term);
        if (posting && posting.documents.includes(proof.documentId)) {
            throw new Error(`term ${proof.term} still linked to document ${proof.documentId}`);
        }
        const group = snapshot.termTombstones.find((entry) => entry.term === proof.term);
        const tombstone = group?.tombstones.find((ts) => ts.documentId === proof.documentId);
        if (!tombstone) {
            throw new Error(`no tombstone for term ${proof.term} and document ${proof.documentId}`);
        }
        const expected = digestFor('term', proof.term, proof.documentId, tombstone.sequence, tombstone.timestamp, tombstone.reason, tombstone.version);
        if (expected !== tombstone.digest) {
            throw new Error(`digest mismatch for term ${proof.term} and document ${proof.documentId}`);
        }
    }
}
